### ISSUE TITLE
eks module adds cluster_security_group_id to eks_managed_node_groups instead of node security group

### ISSUE THREAD

**Issue by youwalther65** [terraform-aws-modules/terraform-aws-eks#1889]
## Description

Using eks_managed_node_groups section in module eks a security group is created called:

$ terraform show
...
    vpc_config                = [
        {
            cluster_security_group_id = "sg-0563e9923f42ba117"

module.eks.aws_security_group.node[0]:
resource "aws_security_group" "node" {
    arn                    = "arn:aws:ec2:eu-west-1:6<redacted:security-group/sg-0xxx"
    description            = "EKS node shared security group"
    egress                 = [

which uses the principle of least priviledge and contains only the necessary ingress and egress rules.
But instead the cluster_security_group_id from vpc_config is attached to the worker nodes from eks_managed_node_groups
## Versions

 - Terraform: v1.1.4

 - Provider(s): hashicorp/aws v4.1.0

 - Module: eks 18.7.0

## Reproduction
module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = ">= 18.7.0"

  eks_managed_node_groups = {
    mng-od = {
      name            = "${local.name}-mng"

## Expected behavior
node security group  module.eks.aws_security_group.node[0] attached to eks_managed_node_groups to implement least priviledge or a properly configured module.eks.module.eks_managed_node_group["mng-od"].aws_security_group.this[0]
Why are they two security groups? 
I have attached the module.eks.aws_security_group.node[0].aws_security-Group manually and it works perfetc.

## Actual behavior
cluster_security_group_id  attached to eks_managed_node_groups
eks_managed_node_group security group has empty ingress and egress rules
module.eks.module.eks_managed_node_group["mng-od"].aws_security_group.this[0]:
resource "aws_security_group" "this" {
    arn                    = "arn:aws:ec2:eu-west-1:<redacted>:security-group/sg-xxx"
    description            = "EKS managed node group security group"
    egress                 = []
    id                     = "sg-xxx"
    ingress                = []
module.eks.aws_security_group.node[0] isn't used (attached) to managed nodes.

**Comment by dejwsz** [comment]
Yes, that is true - I had to add the SG in my node group definition (using submodule) like:

```
  vpc_security_group_ids = [
    module.eks.node_security_group_id
  ]
```

**Comment by Mgonand** [comment]
I'm having this issue too.

**Comment by sly-roar** [comment]
I have faced with the issue as well.

@dejwsz Could you please provide more details about your workaround?
I have tried to do the same
```
eks_managed_node_group_defaults = {
    vpc_security_group_ids = [
      module.eks.node_security_group_id
    ]
  }
```
but did not get success. The cluster security group gets attached to EKS nodes still.
Thank you in advance.

**Comment by dejwsz** [comment]
I have a set of nodes defined using a module and inside it I have node scurity group assignment like (I skipped other details):
```
module "on-demand-set-1" {
....
  vpc_security_group_ids = [
    module.eks.node_security_group_id
  ]
...
}
```
where `node_security_group_id` is exported attribute of previously defined eks module.

**Comment by bryantbiggs** [comment]
using the examples in this project, I am not able to reproduce this issue. On nodes I see two security groups 
1. the shared security group which permits node-to-node communication https://github.com/terraform-aws-modules/terraform-aws-eks/blob/9a99689cc13147f4afc426b34ba009875a28614e/node_groups.tf#L148
2. the security group created by the node group (defaults to no rules) https://github.com/terraform-aws-modules/terraform-aws-eks/blob/9a99689cc13147f4afc426b34ba009875a28614e/modules/eks-managed-node-group/main.tf#L359

without providing a reproduction or clear information, its hard to help support

**Comment by dejwsz** [comment]
Without this explicit security group assignment the nodes defined via module couldn't join the cluster.

**Comment by bryantbiggs** [comment]
> Without this explicit security group assignment the nodes defined via module couldn't join the cluster.

again, you may understand what you are saying but I and others may not because there is no reproduction provided. I have personally provided all of the examples in this project - feel free to copy+paste+update to provide a replica of your setup where you are seeing the issue (reproduction) so that I can better understand what you are referring to

**Comment by sly-roar** [comment]
> I have a set of nodes defined using a module and inside it I have node scurity group assignment like (I skipped other details):
> 
> ```
> module "on-demand-set-1" {
> ....
>   vpc_security_group_ids = [
>     module.eks.node_security_group_id
>   ]
> ...
> }
> ```
> 
> where `node_security_group_id` is exported attribute of previously defined eks module.

Thanks for the clarification! That did the trick.

**Comment by sly-roar** [comment]
> > Without this explicit security group assignment the nodes defined via module couldn't join the cluster.
> 
> again, you may understand what you are saying but I and others may not because there is no reproduction provided. I have personally provided all of the examples in this project - feel free to copy+paste+update to provide a replica of your setup where you are seeing the issue (reproduction) so that I can better understand what you are referring to

I used the approach below for creating EKS cluster and EKS managed nodes:
<details>
  <summary>Spolier</summary>

```
module "eks_label" {
  source      = "cloudposse/label/null"
  version     = "0.25.0"

  namespace   = "example"
  environment = "development"
  attributes  = ["eks"]
  
  tags = {
    Environment = "development"
    Terraform   = "true"
  }
}

module "eks_node_group_label" {
  source      = "cloudposse/label/null"
  version     = "0.25.0"

  namespace   = "example"
  environment = "environment"
  attributes  = ["group"]
  
  tags = {
    Environment = "environment"
    Terraform   = "true"
  }
}

module "eks" {
  source          = "terraform-aws-modules/eks/aws"
  version         = "18.7.2"
  
  cluster_name    = module.eks_label.id
  cluster_version = "1.21"
  cluster_endpoint_private_access = true
  cluster_endpoint_public_access  = true

  cluster_addons = {
    coredns = {
      resolve_conflicts = "OVERWRITE"
    }
    kube-proxy = {}
    vpc-cni = {
      resolve_conflicts = "OVERWRITE"
    }
  }

  cluster_encryption_config = [{
    provider_key_arn = var.k8s_kms_arn
    resources        = ["secrets"]
  }]

  vpc_id = var.vpc_id
  subnet_ids = var.private_subnets

  cloudwatch_log_group_retention_in_days = 7

  enable_irsa = true

  # Extend cluster security group rules
  cluster_security_group_additional_rules = {
    egress_nodes_ephemeral_ports_tcp = {
      description                = "To node 1025-65535"
      protocol                   = "tcp"
      from_port                  = 1025
      to_port                    = 65535
      type                       = "egress"
      source_node_security_group = true
    }
  }

  # Extend node-to-node security group rules
  node_security_group_additional_rules = {
    ingress_self_all = {
      description = "Node to node all ports/protocols"
      protocol    = "-1"
      from_port   = 0
      to_port     = 0
      type        = "ingress"
      self        = true
    }
    ingress_lb_all = {
      description = "Load Balancer to node all protocols port 30100"
      protocol    = "tcp"
      from_port   = 30100
      to_port     = 30100
      type        = "ingress"
      cidr_blocks = var.public_subnets_cidr_blocks
    }
    egress_all = {
      description      = "Node all egress"
      protocol         = "-1"
      from_port        = 0
      to_port          = 0
      type             = "egress"
      cidr_blocks      = ["0.0.0.0/0"]
      ipv6_cidr_blocks = ["::/0"]
    }
  }

  eks_managed_node_group_defaults = {
    force_update_version  = true
    ami_type              = "BOTTLEROCKET_x86_64"
    platform              = "bottlerocket"
    disk_size             = 30
    ebs_optimized         = true
  }

  eks_managed_node_groups = {
    develop = {
      # By default, the module creates a launch template to ensure tags are propagated to instances, etc.,
      # so we need to disable it to use the default template provided by the AWS EKS managed node group service
      create_launch_template = false
      launch_template_name   = ""

      instance_types                = ["t3.small"]
      capacity_type                 = "SPOT"
      min_size                      = 1
      max_size                      = 3
      desired_size                  = 2

      metadata_options = {
        http_endpoint               = "enabled"
        http_tokens                 = "required"
        http_put_response_hop_limit = 2
        instance_metadata_tags      = "disabled"
      }

      labels = {
        Environment = "development"
      }
      
      tags = module.eks_node_group_label.tags
    }
  }
}
  ```
</details>
Hope this helps in reproducing the issue.

**Comment by alt-dima** [comment]
To reproduce it, specify in the eks_managed_node_groups.
```
create_launch_template = false
aunch_template_name   = ""
```
Example:

```
    eks_managed_node_groups = {
        kconsumers = {
            name = "test"

            create_launch_template = false
            launch_template_name   = ""

            min_size     = 1
            max_size     = 1
            desired_size = 1

            instance_types = ["r5.large"]
            capacity_type  = "SPOT"
            disk_size = 100

        }
```

**Comment by bryantbiggs** [comment]
In this instance, setting
```hcl
  create_launch_template = false
  launch_template_name   = ""
```

Means you are relying on the default launch template provided by the EKS managed node group service and therefore custom settings like security groups (and tags, etc.) are not available. This module has made the decision to default to using a custom launch template since these are settings users typically expect and rely on. 

tl;dr - yes, using the default EKS managed node group service will NOT use the node security group and instead attach the cluster security group, just like when running a Fargate profile
